cpp/src/util/main/UnitTests.cpp

```diff
@@ -125,6 +125,343 @@ TEST(eigen_util, sort_columns) {
+  }
+}
+
+TEST(eigen_util, softmax_Tensor) {
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<N, 1>>;
+
+  Tensor tensor;
+  tensor.setValues({{0}, {1}, {2}, {3}});
+  Tensor expected;
+  expected.setValues({{0.0320586}, {0.0871443}, {0.2368828}, {0.6439143}});
+
+  tensor = eigen_util::softmax<Tensor>(tensor);
+
+  for (int i = 0; i < N; ++i) {
+    EXPECT_NEAR(tensor(i, 0), expected(i, 0), 1e-5);
+  }
+}
+
+TEST(eigen_util, sigmoid) {
+  constexpr int N = 4;
+  using Array = eigen_util::FArray<N>;
+
+  Array array{{0, 1, 2, 3}};
+  Array expected{{0.5, 0.7310586, 0.8807971, 0.9525741}};
+
+  array = eigen_util::sigmoid<Array>(array);
+
+  for (int i = 0; i < N; ++i) {
+    EXPECT_NEAR(array(i), expected(i), 1e-5);
+  }
+}
+
+TEST(eigen_util, reverse) {
+  constexpr int M = 3;
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<3, 4>>;
+
+  Tensor tensor;
+  tensor.setValues({{0, 1, 2, 3}, {4, 5, 6, 7}, {8, 9, 10, 11}});
+  Tensor expected;
+  expected.setValues({{3, 2, 1, 0}, {7, 6, 5, 4}, {11, 10, 9, 8}});
+
+  Tensor reversed = eigen_util::reverse<Tensor>(tensor, 1);
+
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      EXPECT_EQ(reversed(i, j), expected(i, j));
+    }
+  }
+}
+
+TEST(eigen_util, normalize) {
+  constexpr int M = 2;
+  constexpr int N = 4;
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<M, N>>;
+  using Array = Eigen::Array<float, M, N, Eigen::RowMajor>;
+
+  Array array{{0, 1, 2, 3}, {0, 0, 4, 0}};
+  Array expected = array / array.sum();
+  Tensor tensor = Eigen::TensorMap<Tensor>(array.data(), array.rows(), array.cols());
+
+  bool success = eigen_util::normalize<Tensor>(tensor, 1e-5);
+
+  EXPECT_TRUE(success);
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      EXPECT_NEAR(tensor(i, j), expected(i, j), 1e-5);
+    }
+  }
+}
+
+TEST(eigen_util, sample) {
+  constexpr int N = 4;
+  constexpr int numSamples = 10000;
+
+  using Array = eigen_util::FArray<N>;
+
+  Array values = {{1, 2, 3, 4}};
+  Array expectedFreq = values / values.sum() * numSamples;
+
+  Array freq;
+  freq.setZero();
+  for (int i = 0; i < numSamples; i++) {
+    auto sample = eigen_util::sample<Array>(values);
+    freq(sample)++;
+  }
+
+  // Chi-Squared Statistic \sum{(O_i - E_i)^2 / E_i}
+  // where O_i is the observed frequency and E_i is the expected frequency
+  float chi2 = 0;
+  for (int i = 0; i < N; i++) {
+    float diff = freq(i) - expectedFreq(i);
+    chi2 += diff * diff / expectedFreq(i);
+  }
+
+  // The chi-squared value corresponding to a p-value of 0.05 with 3 degrees of freedom is 7.81
+  EXPECT_LT(chi2, 7.81);
+}
+
+TEST(eigen_util, randomly_zero_out) {
+  constexpr int M = 8;
+  constexpr int N = 4;
+  constexpr int numZeroes = 10;
+  static_assert(numZeroes <= M * N, "numZeroes must be less than or equal to M * N");
+
+  using Tensor = eigen_util::FTensor<Eigen::Sizes<M, N>>;
+  Tensor tensor;
+  tensor.setRandom();
```

Comment:
This `setRandom()` call uses a random-source from within `eigen3`, which will *not* be the same random-source that is used by `eigen_util` (which gets its randomness from `util::Random`). The `eigen3` library undoubtedly seeds its RNG with something non-static (such as current-time) by default. Thus, this will result in a different `tensor` every time this test is run.

Theoretically, although it's **highly** unlikely, one of the tensor elements could randomly get assigned a value of exactly zero. That could cause this test to fail unexpectedly.

There is no reason in this test to set `tensor` randomly. Instead, you should do something like `tensor.setConstant(1)`.

Reply:
Understood. Here I used EXPECT_GE to account for the zero element scenario but also thought it was not ideal. Yeah will update it to constant.

Reply:
Oh I see. Yes, with `EXPECT_GE`, your test should always pass. But it's still better to avoid this unnecessary randomness.